java/arcs/core/util/Parser.kt

```diff
@@ -0,0 +1,278 @@
+    fun <S> flatMap(f: (T, String) -> ParseResult<S>): ParseResult<S> {
+        return when (this) {
+            is Success<T> -> f(value, rest)
+            else -> Failure
+        }
+    }
+
+    /** Represents a successful parse, containing the parsed value, and unparsed leftover. */
+    data class Success<out T>(val value: T, val rest: String) : ParseResult<T>()
+
+    /** Represents a parse failure. */
+    object Failure : ParseResult<Nothing>()
+}
+
+/** Chop off the consumed part of the string. */
+fun String.advance(str: String) = this.substring(str.length)
+
+/** A parser that consumes a prefix of a string. */
+class StringToken(val token: String) : Parser<String>() {
+    override fun invoke(string: String): ParseResult<String> = when {
+        string.startsWith(token) -> Success(token, string.advance(token))
+        else -> Failure
+    }
+}
+
+/**
+ * A parser that consumes a regex prefix (anchored at ^) of a String, and returns the
+ * first matchgroup.
+ */
+class RegexToken(val regexToken: String) : Parser<String>() {
+    override fun invoke(string: String): ParseResult<String> =
+        Regex("^${regexToken}").find(string, 0)?.let { it ->
+            Success(it.groupValues[1], string.advance(it.groupValues[0]))
+        } ?: Failure
+}
+
+/**
+ * A parser that invokes another parser, and if it success, it returns, otherwise is succeeds with
+ * null.
+ */
+class Optional<T>(val parser: Parser<T>) : Parser<T?>() {
+    override fun invoke(string: String): ParseResult<T?> = parser(string).orElse {
+        Success<T?>(null, string)
+    }
+}
+
+/**
+ * A parser that combines two parsers in sequence, if either one fails, the parse fails,
+ * otherwise the combined results are returned as a [Pair<T, S].
+ */
+class Seq<T, S>(val left: Parser<T>, val right: Parser<S>) : Parser<Pair<T, S>>() {
+    override fun invoke(string: String) = left(string).flatMap { v1, r ->
+        right(r).flatMap { v2, r2 ->
+            Success(Pair(v1, v2), r2)
+        }
+    }
+}
+
+/** A parser that combines two parsers by returning the value of the first one that succeeds. */
+class Par<T>(val left: Parser<T>, val right: Parser<T>) : Parser<T>() {
+    override fun invoke(string: String): ParseResult<T> = left(string).orElse {
+        right(string)
+    }
+}
+
+/**
+ * A parser that invokes another parser zero or more times until it fails, returning a list
+ * of success values. If no parse succeeds, it returns an empty list.
+ */
+class Many<T>(val parser: Parser<T>) : Parser<List<T>>() {
+    override fun invoke(string: String): ParseResult<List<T>> {
+        val result = mutableListOf<T>()
+        var parseResult = parser(string)
+        var lastSuccess = parseResult
+        while (parseResult is Success<T>) {
+            result.add(parseResult.value)
+            lastSuccess = parseResult
+            parseResult = parser(parseResult.rest)
+        }
+        return Success(result, (lastSuccess as? Success<T>)?.rest ?: string)
```

Comment:
Maybe something like this (may be invalid Kotlin) would be cleaner?
```
        val result = mutableListOf<T>()
        var rest = string
        do {
            when (parser(rest)) {
               is Success<T> -> {
                  result.add(parseResult.value)
                  rest = parseResult.rest
               }
               is Failure -> break
            }
        } while (true)
        return Success(result, rest)
```

Reply:
I went rogue and rewrote it functional style :)